File: api/update-interface-name.js
const { default: fetch } = require('node-fetch');

module.exports = async (req, res) => {
  const functionAppUrl = process.env.AZURE_FUNCTION_APP_URL;
  
  if (!functionAppUrl) {
    return res.status(500).json({ 
      error: 'Please check Function App configuration and ensure AZURE_FUNCTION_APP_URL is set',
      url: 'Not configured'
    });
  }

  try {
    const { oldInterfaceName, newInterfaceName } = req.body;

    if (!oldInterfaceName || !newInterfaceName) {
      return res.status(400).json({ 
        error: 'oldInterfaceName and newInterfaceName are required',
        url: `${functionAppUrl}/api/UpdateInterfaceName`
      });
    }

    const functionUrl = `${functionAppUrl}/api/UpdateInterfaceName?code=${process.env.AZURE_FUNCTION_KEY || ''}`;
    
    const response = await fetch(functionUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        oldInterfaceName,
        newInterfaceName
      })
    });

    if (!response.ok) {
      const errorText = await response.text();
      let errorData;
      try {
        errorData = JSON.parse(errorText);
      } catch {
        errorData = { error: errorText };
      }
      
      return res.status(response.status).json({
        ...errorData,
        url: functionUrl,
        status: response.status,
        statusText: response.statusText
      });
    }

    const data = await response.json();
    res.json(data);
  } catch (error) {
    console.error('Error updating interface name:', error);
    res.status(500).json({ 
      error: error.message || 'An error occurred while updating interface name',
      url: `${functionAppUrl}/api/UpdateInterfaceName`,
      details: error.stack
    });
  }
};

File: azure-functions/main/UpdateInterfaceName.cs
using System.Text.Json;
using Microsoft.Azure.Functions.Worker;
using Microsoft.Azure.Functions.Worker.Http;
using Microsoft.Extensions.Logging;
using InterfaceConfigurator.Main.Core.Interfaces;
using InterfaceConfigurator.Main.Helpers;

namespace InterfaceConfigurator.Main;

public class UpdateInterfaceName
{
    private readonly IInterfaceConfigurationService _configService;
    private readonly ILogger<UpdateInterfaceName> _logger;

    public UpdateInterfaceName(
        IInterfaceConfigurationService configService,
        ILogger<UpdateInterfaceName> logger)
    {
        _configService = configService ?? throw new ArgumentNullException(nameof(configService));
        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
    }

    [Function("UpdateInterfaceName")]
    public async Task<HttpResponseData> Run(
        [HttpTrigger(AuthorizationLevel.Anonymous, "post", "options", Route = "UpdateInterfaceName")] HttpRequestData req,
        FunctionContext executionContext)
    {
        _logger.LogInformation("UpdateInterfaceName function triggered");

        // Handle CORS preflight requests
        if (req.Method.Equals("OPTIONS", StringComparison.OrdinalIgnoreCase))
        {
            var optionsResponse = req.CreateResponse(System.Net.HttpStatusCode.OK);
            CorsHelper.AddCorsHeaders(optionsResponse);
            return optionsResponse;
        }

        try
        {
            var requestBody = await new StreamReader(req.Body).ReadToEndAsync();
            var request = JsonSerializer.Deserialize<UpdateInterfaceNameRequest>(requestBody);

            if (request == null || string.IsNullOrWhiteSpace(request.OldInterfaceName) || string.IsNullOrWhiteSpace(request.NewInterfaceName))
            {
                var badRequestResponse = req.CreateResponse(System.Net.HttpStatusCode.BadRequest);
                badRequestResponse.Headers.Add("Content-Type", "application/json; charset=utf-8");
                CorsHelper.AddCorsHeaders(badRequestResponse);
                await badRequestResponse.WriteStringAsync(JsonSerializer.Serialize(new { error = "OldInterfaceName and NewInterfaceName are required" }));
                return badRequestResponse;
            }

            // Get existing configuration
            var config = await _configService.GetConfigurationAsync(request.OldInterfaceName, executionContext.CancellationToken);
            if (config == null)
            {
                var notFoundResponse = req.CreateResponse(System.Net.HttpStatusCode.NotFound);
                notFoundResponse.Headers.Add("Content-Type", "application/json; charset=utf-8");
                CorsHelper.AddCorsHeaders(notFoundResponse);
                await notFoundResponse.WriteStringAsync(JsonSerializer.Serialize(new { error = $"Interface configuration '{request.OldInterfaceName}' not found" }));
                return notFoundResponse;
            }

            // Check if new name already exists
            var existingWithNewName = await _configService.GetConfigurationAsync(request.NewInterfaceName, executionContext.CancellationToken);
            if (existingWithNewName != null && existingWithNewName.InterfaceName != request.OldInterfaceName)
            {
                var conflictResponse = req.CreateResponse(System.Net.HttpStatusCode.Conflict);
                conflictResponse.Headers.Add("Content-Type", "application/json; charset=utf-8");
                CorsHelper.AddCorsHeaders(conflictResponse);
                await conflictResponse.WriteStringAsync(JsonSerializer.Serialize(new { error = $"Interface configuration '{request.NewInterfaceName}' already exists" }));
                return conflictResponse;
            }

            // Update interface name using service method
            await _configService.UpdateInterfaceNameAsync(request.OldInterfaceName, request.NewInterfaceName, executionContext.CancellationToken);

            // Get updated configuration
            config = await _configService.GetConfigurationAsync(request.NewInterfaceName, executionContext.CancellationToken);

            _logger.LogInformation("Updated interface name from '{OldName}' to '{NewName}'", request.OldInterfaceName, request.NewInterfaceName);

            var response = req.CreateResponse(System.Net.HttpStatusCode.OK);
            response.Headers.Add("Content-Type", "application/json; charset=utf-8");
            CorsHelper.AddCorsHeaders(response);
            await response.WriteStringAsync(JsonSerializer.Serialize(config));
            return response;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error updating interface name");
            var errorResponse = req.CreateResponse(System.Net.HttpStatusCode.InternalServerError);
            errorResponse.Headers.Add("Content-Type", "application/json; charset=utf-8");
            CorsHelper.AddCorsHeaders(errorResponse);
            await errorResponse.WriteStringAsync(JsonSerializer.Serialize(new { error = ex.Message }));
            return errorResponse;
        }
    }

    private class UpdateInterfaceNameRequest
    {
        public string OldInterfaceName { get; set; } = string.Empty;
        public string NewInterfaceName { get; set; } = string.Empty;
    }
}




